src/sentinel.rs

```diff
@@ -82,44 +73,41 @@ impl<'a, Request, Name>
     /// signing key. Each such a public signing key needs keys_threshold confirmations
     /// for it to be considered valid and used for verifying the signature
     /// of the corresponding claim.
-    pub fn new(sender: &'a mut GetSigningKeys<Name>, claim_threshold: usize, keys_threshold: usize)
-        -> Sentinel<'a, Request, Name> {
+    pub fn new(claim_threshold: usize, keys_threshold: usize)
+        -> Sentinel<Request, Name> {
         Sentinel {
-            sender: sender,
             claim_accumulator: Accumulator::new(claim_threshold),
-            keys_accumulator: Accumulator::new(keys_threshold),
+            key_store: KeyStore::<Name>::new(keys_threshold),
             claim_threshold: claim_threshold,
-            keys_threshold: keys_threshold
         }
     }
 
-    /// This adds a new claim for the provided request.
-    /// The name and the signature provided will be used to validate the claim
-    /// with the keys that are independently retrieved.
-    /// When an added claim leads to the resolution of the request,
-    /// the request and the claim are returned.
-    /// All resolved claims have to be identical.
-    /// Otherwise None is returned.
-    pub fn add_claim(&mut self, request : Request,
-                     claimant : Name, signature : Signature,
-                     claim : SerialisedClaim)
-        // TODO: replace return option with async events pipe to different thread
-        // TODO: code can be cleaned up more even by correcting ownership of Accumulator
-        -> Option<(Request, SerialisedClaim)> {
-
-        match self.keys_accumulator.get(&request) {
-            Some((_, set_of_keys)) => {
-                self.claim_accumulator.add(request.clone(), (claimant, signature, claim))
-                    .and_then(|(_, claims)| self.validate(&claims, &set_of_keys))
-                    .and_then(|verified_claims| self.resolve(&verified_claims))
-                    .map(|merged_claim| (request, merged_claim))
-            },
-            None => {
-                self.sender.get_signing_keys(request.get_source());
-                self.claim_accumulator.add(request, (claimant, signature, claim));
-                return None;
+    /// This adds a new claim for the provided request. The claimant name and
+    /// the signature provided will be used to verify the claim with the keys
+    /// that are independently retrieved. When an added claim leads to the
+    /// resolution of the request, the request and the claim are returned.
+    /// All resolved claims have to be identical. Otherwise None is returned.
+    pub fn add_claim(&mut self,
```

Comment:
Maybe mention in the documentation also that RequestKeys(Name) is returned when Sentinel requires the keys